x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/InternalExecutePolicyAction.java

```diff
@@ -0,0 +1,170 @@
+        private final TransportService transportService;
+        private final EnrichPolicyExecutor policyExecutor;
+        private final AtomicInteger nodeGenerator = new AtomicInteger(Randomness.get().nextInt());
+
+        @Inject
+        public Transport(
+            TransportService transportService,
+            ActionFilters actionFilters,
+            ClusterService clusterService,
+            EnrichPolicyExecutor policyExecutor
+        ) {
+            super(NAME, transportService, actionFilters, Request::new);
+            this.clusterService = clusterService;
+            this.transportService = transportService;
+            this.policyExecutor = policyExecutor;
+        }
+
+        @Override
+        protected void doExecute(Task transportTask, Request request, ActionListener<Response> actionListener) {
+            var clusterState = clusterService.state();
+            var node = selectNodeForPolicyExecution(clusterState.nodes());
+            if (clusterState.nodes().getLocalNode().equals(node) == false) {
+                var handler = new ActionListenerResponseHandler<>(actionListener, Response::new);
+                transportService.sendRequest(node, NAME, request, handler);
+                return;
+            }
+
+            // Can't use provided task, because in the case wait_for_completion=false then
+            // as soon as actionListener#onResponse is invoked then the provided task get unregistered and
+            // then there no way to see the policy execution in the list tasks or get task APIs.
+            var task = (ExecuteEnrichPolicyTask) taskManager.register("enrich", TASK_ACTION, new TaskAwareRequest() {
+
+                @Override
+                public void setParentTask(TaskId taskId) {
+                    request.setParentTask(taskId);
+                }
+
+                @Override
+                public TaskId getParentTask() {
+                    return request.getParentTask();
+                }
+
+                @Override
+                public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+                    String description = "executing enrich policy [" + request.getName() + "]";
+                    return new ExecuteEnrichPolicyTask(id, type, action, description, parentTaskId, headers);
+                }
+            });
+
+            try {
+                ActionListener<ExecuteEnrichPolicyStatus> listener;
+                if (request.isWaitForCompletion()) {
+                    listener = ActionListener.wrap(result -> actionListener.onResponse(new Response(result)), actionListener::onFailure);
+                } else {
+                    listener = ActionListener.wrap(result -> {
+                        LOGGER.debug("successfully executed policy [{}]", request.getName());
+                        // taskManager.storeResult(task, new Response(result), ActionListener.wrap(() -> {/*noop*/}));
```

Comment:
I will remove this. 

I was playing around with storing the result of policy execution in the `.tasks` index. So that we can consistently report on async executed policy executions. Do you think we should do that in a followup change?

Reply:
I think that's a good idea. I only see a couple of places where we actually store task data after it completes. Seems like Reindex, and a couple of ML actions. I don't see any reason we shouldn't be storing them, though IIRC the user needs to clear them out themselves over time...

Reply:
Cool, I will add this to #48988

>  though IIRC the user needs to clear them out themselves over time...

True, but I don't expect this index will ever contain a massive amount of data. Ideally this should be a data stream, so that ilm can manage it.

Reply:
Removed commented out code: 3e455d41ab5f29bdbf70c989aa7b3291eb62735a